src/kde/kde.cc

```diff
@@ -30,124 +30,437 @@ namespace rtt_kde {
+  double reconstruction_conservation =
+      std::accumulate(new_distribution.begin(), new_distribution.end(), 0.0);
+  double abs_distribution_conservation =
+      std::accumulate(abs_distribution.begin(), abs_distribution.end(), 0.0);
 
   if (domain_decomposed) {
     // accumulate global contribution
+    rtt_c4::global_sum(original_conservation);
     rtt_c4::global_sum(reconstruction_conservation);
+    rtt_c4::global_sum(abs_distribution_conservation);
   }
 
-  if (!rtt_dsxx::soft_equiv(reconstruction_conservation, 0.0) &&
-      !rtt_dsxx::soft_equiv(global_conservation, 0.0)) {
-    // Totals are non-zero so scale the result for conservation
-    for (int i = 0; i < local_size; i++)
-      result[i] *= global_conservation / reconstruction_conservation;
-  } else {
-    // a zero distribution is possible. If it occurs fall back to residual conservation;
-    const double res = global_conservation - reconstruction_conservation;
-    for (int i = 0; i < local_size; i++)
-      result[i] += res / double(size);
+  // Apply residual
+  if (abs_distribution_conservation > 0.0) {
+    const double res = original_conservation - reconstruction_conservation;
+    for (size_t i = 0; i < local_size; i++)
+      new_distribution[i] += res * abs_distribution[i] / abs_distribution_conservation;
   }
+}
 
-  return result;
+//------------------------------------------------------------------------------------------------//
+/*!
+ * \brief 
+ *
+ *  Calculate the bounding window (via win_min (x_min,y_min,z_min) and win_max
+ *  (x_max, y_max, z_max)) given a central location and the bandwidth size in
+ *  each dimension (dx,dy) for Cartesian or (dr,arc_length) for spherical.
+ * 
+ * \param[in] qindex quick index class for finding bounds xy bounds of a wedge shape
+ * \param[in] position is the central location of the bounds
+ * \param[in] one_over_bandwidth size of the reconstruction domain in each dimension. This is (dx,dy) for Caresian and (dr, arc_length) for spherical. 
+ * \param[in,out] win_min is the minimum corner of the bounding box (x_min, y_min, z_min)
+ * \param[in,out] win_max is the maximum corner of the bounding box (x_max, y_max, z_max)
+ *
+ */
+void kde::calc_win_min_max(const quick_index &qindex, const std::array<double, 3> &position,
+                           const std::array<double, 3> &one_over_bandwidth,
+                           std::array<double, 3> &win_min, std::array<double, 3> &win_max) const {
+  size_t dim = qindex.dim;
+  if (use_spherical_reconstruction) {
+    const double dr = 1.0 / one_over_bandwidth[0];
+    const double rmax = sqrt((sphere_center[0] - position[0]) * (sphere_center[0] - position[0]) +
+                             (sphere_center[1] - position[1]) * (sphere_center[1] - position[1])) +
+                        dr;
+    // dtheta = arch_length_max/rmax
+    const double dtheta = std::min(1.0 / (one_over_bandwidth[1] * rmax), acos(0));
```

Comment:
Can this be a division-by-zero?

Reply:
Covered by new require and check statements.